cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaReconciler.java

```diff
@@ -1064,4 +1124,43 @@ protected Future<Void> updateNodePoolStatuses(KafkaStatus kafkaStatus) {
             return Future.succeededFuture();
         }
     }
+
+    /**
+     * Validate the Kafka version set in the Kafka custom resource, together with the metadata version and the configured
+     * inter.broker.protocol.version and log.message.format.version.
+     * They need to be all aligned and at least 3.4 to support ZooKeeper to KRaft migration.
+     *
+     * @param kafkaStatus Status of the Kafka custom resource where warnings about any issues with metadata state will be added
+     *
+     * @return Future which completes when validation is done
+     */
+    Future<Void> validateVersionsForKRaftMigration(KafkaStatus kafkaStatus) {
```

Comment:
All the information I need for the validation are coming from the `KafkaCluster` instance (i.e. Kafka version, metadata version, IBP and LMF). While I can see a way to get Kafka version, IBP and LMF from the Kafka CR (which is anyway by duplicating code which is happening in the `KafkaCluster` class), I can't see a good way to get metadata version.
It's coming trough the `KafkaVersionChange` which, in the KAO, is created later and after all the validations when it runs the multiple compose(s) in the reconcile. Is there any other way I am missing to get the metadata version from the Kafka CR? Otherwise moving this metod out of `KafkaReconciler` makes things really complicated and not sure it's worth it.

Reply:
As agreed offline I moved this validation method in the `KRaftUtils` and the validation now runs in the `KafkaCluster`.